### PR TITLE
Fix many remaining bugs with >2gb address space

### PR DESCRIPTION
**Pull request by sbc100**

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -903,6 +907,9 @@ workflows:
       - test-wasm2js1:
           requires:
             - build-linux
+      - test-core_2gb:
+          requires:
+            - build-linux
```
Comment (kripken) [inline .circleci/config.yml:912]: I am a little unsure this is worth it. It won't slow the time to get the final results, but it does add to the total budget of work.

Isn't this option mostly useful for us during development? I was imagining we'd run it locally.

File: emcc.py [hunk 2]
```diff
@@ -1781,10 +1781,14 @@ def set_max_memory():
       diagnostics.warning('unused-command-line-argument', 'MAXIMUM_MEMORY is only meaningful with ALLOW_MEMORY_GROWTH')
     settings.MAXIMUM_MEMORY = settings.INITIAL_MEMORY
 
-  # INITIAL_MEMORY sets a lower bound for MAXIMUM_MEMORY
   if 'MAXIMUM_MEMORY' not in user_settings:
-    if settings.INITIAL_MEMORY > settings.MAXIMUM_MEMORY:
-      settings.MAXIMUM_MEMORY = settings.INITIAL_MEMORY
+    if settings.ALLOW_MEMORY_GROWTH:
+      if settings.INITIAL_MEMORY > 2 * 1024 * 1024 * 1024:
+        settings.MAXIMUM_MEMORY = 4 * 1024 * 1024 * 1024
```
Comment (kripken) [inline emcc.py:1787]: It seems like this disallows an initial size of 2.01 GB that is growable up to 3 GB (but no more)?

File: test/common.py [hunk 3]
```diff
@@ -339,8 +339,11 @@ def metafunc(self, with_wasm64):
 def can_do_standalone(self, impure=False):
   # Pure standalone engines don't support MEMORY64 yet.  Even with MEMORY64=2 (lowered)
   # the WASI APIs that take pointer values don't have 64-bit variants yet.
-  if self.get_setting('MEMORY64') and not impure:
-    return False
+  if not impure:
+    if self.get_setting('MEMORY64'):
+      return False
+    if self.get_setting('INITIAL_MEMORY') == '2200mb':
```
Comment (kripken) [inline test/common.py:345]: This magic number is surely worth a comment :smile:

File: test/core/test_alloca.out [hunk 4]
```diff
@@ -1 +1 @@
-z:1*
\ No newline at end of file
+
```
Comment (kripken) [inline test/core/test_alloca.out:1]: Do we need this file if it's empty?

File: test/test_core.py [hunk 5]
```diff
@@ -266,6 +266,21 @@ def decorated(self, *args, **kwargs):
...
 
+def no_2gb(note):
+  assert not callable(note)
+
+  def decorator(f):
+    assert callable(f)
+
+    @wraps(f)
+    def decorated(self, *args, **kwargs):
+      if self.get_setting('INITIAL_MEMORY') == '2200mb':
```
Comment (kripken) [inline test/test_core.py:277]: Also here.

File: test/test_core.py [hunk 6]
```diff
@@ -2757,7 +2783,8 @@ def test_pthread_proxying(self, modularize):
   @node_pthreads
   def test_pthread_proxying_cpp(self):
     self.set_setting('PROXY_TO_PTHREAD')
-    self.set_setting('INITIAL_MEMORY=32mb')
+    if not self.has_changed_setting('INITIAL_MEMORY'):
+      self.set_setting('INITIAL_MEMORY=32mb')
```
Comment (kripken) [inline test/test_core.py:2787]: Maybe worth a function that changes a setting to a value if it hasn't been changed?  I.e., sets a new "default". (doesn't need to be in this PR)

File: emcc.py (hunk 2 above)
Comment (sbc100) [reply emcc.py:1787]: This entire block only runs if the user doesn't specify their own `MAXIMUM_MEMORY`.

File: test/core/test_alloca.out (hunk 4 above)
Comment (sbc100) [reply test/core/test_alloca.out:1]: There is no useful output from this test.. so I guess I can convert the type of test to one that doesn't expect an output file and delete this.

File: emcc.py (hunk 2 above)
Comment (kripken) [reply emcc.py:1787]: I see, thanks.

File: .circleci/config.yml (hunk 1 above)
Comment (sbc100) [reply .circleci/config.yml:912]: I would worry that we could then be in danger of introducing regressions, but on the other hand, we do have many test modes that we don't run in CI. 

I will remove for now, once I see it passing.

Comment (sbc100) [reply .circleci/config.yml:912]: > Another option would be to run it on the testsuite bot on Chromium CI.

Yes, lets do that.

**Comment by sbc100** [comment]
OK, I reverted the addition of the new test config in circleCI.  PTAL.

**Review by kripken — APPROVED** [review]

**Pull request merged**